### PR TITLE
Fix push_down_projection through a distinct

### LINKED ISSUE #4667: `SELECT ... FROM (tbl1 UNION tbl2)` wrongly works like `SELECT DISTINCT ... FROM (tbl1 UNION tbl2)`
The title says it all. But in other words: 
if a table constructed with `UNION [DISTINCT]` is used as a subtable for a `SELECT` statement,
and not all columns of the table are mentioned in the select, the output is incorrectly 
deduplicated while considering only the mentioned columns, instead of all columns of the subtable.

## To Reproduce
Consider a simple table, constructed with a UNION statement:
```
+----+------+
| id | data |
+----+------+
| 1  | 10   |
| 1  | 20   |
+----+------+
```
Let's select the `id` column from it:
```rust
let ctx = SessionContext::new();
let df = ctx.sql("SELECT id FROM (SELECT 1 as id, 10 as data UNION SELECT 1, 20)").await?;
df.show().await?;
```
## Expected output
```
+----+
| id |
+----+
| 1  |
| 1  |
+----+
```
## Actual output
```
+----+
| id |
+----+
| 1  |
+----+
```
## Additional context
Of course, in this particular example this wrong behavior could be worked around by using `UNION ALL`, but this isn't always possible.
E.g. the following query, which detects rows that have same ids but differing data in other columns, can't be (easily) rewritten in terms of `UNION ALL`:
```rust
let df = ctx.sql("SELECT id FROM (SELECT * FROM tbl1 UNION SELECT * FROM tbl2) GROUP BY id HAVING count(*) = 2").await?;
```

### PR DESCRIPTION
**Pull request by Jefffrey**
# Which issue does this PR close?

Closes #4667

# Rationale for this change

During PushDownProjection optimization, it does not properly account for distinct, as it will attempt to push directly through the distinct, from input plan:

```sql
Projection: data.id
  Distinct:
    Projection: data.id, data.data
      TableScan: data
```

Will generate output plan:

```sql
Projection: data.id
  Distinct:
    TableScan: data projection=[id]
```

Where it has inadvertently removed the `data.data` projection during the push down and hence leading to incorrect distinct behaviour.

# What changes are included in this PR?

Properly handle distincts during recursion in PushDownProjection, to 'restart' the traversal with all required columns (derived from schema of the distinct) rather than using the input required columns.

After fix, the above plan output will instead be:

```sql
Projection: data.id
  Distinct:
    TableScan: data projection=[id, data]
```

# Are these changes tested?

New unit test in push_down_projection

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/optimizer/src/push_down_projection.rs [hunk 1]
```diff
@@ -374,6 +374,24 @@ fn optimize_plan(
             )?;
             from_plan(plan, &plan.expressions(), &[child])
         }
+        // at a distinct, all columns are required
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_projection.rs:377]: 👍

File: datafusion/optimizer/src/push_down_projection.rs [hunk 2]
```diff
@@ -1009,6 +1026,25 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn pushdown_through_distinct() -> Result<()> {
+        let table_scan = test_table_scan()?;
+
+        let plan = LogicalPlanBuilder::from(table_scan)
+            .project(vec![col("a"), col("b")])?
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_projection.rs:1034]: 👍

**Review by alamb — APPROVED** [review]
LGTM -- thank you @Jefffrey

**Comment by alamb** [comment]
cc @jackwener

**Review by jackwener — APPROVED** [review]
Make sense to me, thanks @Jefffrey

**Pull request merged**